docs/content/guides/dagster/enriching-with-software-defined-assets.mdx

```diff
@@ -0,0 +1,407 @@
+
+#### Without software-defined assets
+
+This example is a vanilla, op-based job that follows the idiomatic practice of delegating all IO to IO managers and root input managers.
+
+The goal of each op in the job is to produce an asset. However, because the job doesn't use the software-defined asset APIs, Dagster is unaware of this:
+
+```python file=/guides/dagster/enriching_with_software_defined_assets/vanilla_io_manager.py
+from pandas import DataFrame
+
+from dagster import In, Out, job, op, repository
+
+from .mylib import s3_io_manager, snowflake_io_manager, train_recommender_model
+
+
+@op(
+    ins={"raw_users": In(root_manager_key="warehouse")},
+    out={"users": Out(io_manager_key="warehouse")},
+)
+def build_users(raw_users: DataFrame) -> DataFrame:
+    users_df = raw_users.dropna()
+    return users_df
+
+
+@op(out={"users_recommender_model": Out(io_manager_key="object_store")})
+def build_user_recommender_model(users: DataFrame):
+    users_recommender_model = train_recommender_model(users)
+    return users_recommender_model
+
+
+@job(resource_defs={"warehouse": snowflake_io_manager, "object_store": s3_io_manager})
+def users_recommender_job():
+    build_user_recommender_model(build_users())
+
+
+@repository
+def repo():
+    return [users_recommender_job]
+```
+
+#### With software-defined assets
+
+Here's what an equivalent job looks like using software-defined assets:
+
+```python file=/guides/dagster/enriching_with_software_defined_assets/sda_io_manager.py
+from pandas import DataFrame
+
+from dagster import SourceAsset, asset, define_asset_job, repository, with_resources
+
+from .mylib import s3_io_manager, snowflake_io_manager, train_recommender_model
+
+raw_users = SourceAsset(key="raw_users", io_manager_key="warehouse")
+
+
+@asset(io_manager_key="warehouse")
+def users(raw_users: DataFrame) -> DataFrame:
+    users_df = raw_users.dropna()
+    return users_df
+
+
+@asset(io_manager_key="object_store")
+def user_recommender_model(users: DataFrame):
+    users_recommender_model = train_recommender_model(users)
+    return users_recommender_model
+
+
+@repository
+def repo():
+    return [
+        *with_resources(
+            [raw_users, users, user_recommender_model],
+            resource_defs={
+                "warehouse": snowflake_io_manager,
+                "object_store": s3_io_manager,
+            },
+        ),
+        define_asset_job("users_recommender_job"),
+    ]
+```
+
+### Materialize two interdependent tables without an IO manager
+
+- [Without software-defined-assets](#without-software-defined-assets-1)
+- [With software-defined assets](#with-software-defined-assets-1)
+
+#### Without software-defined assets
+
+This example does the same things as the [previous example](#a-job-that-materializes-two-interdependent-tables), with one difference. This job performs IO inside of the ops instead of delegating it to IO managers and root input managers:
```

Comment:
```suggestion
This example does the same things as the [previous example](#materialize-two-interdependent-tables), with one difference. This job performs IO inside of the ops instead of delegating it to IO managers and root input managers:
```